[1단계 - 자동차 경주] 분주(황현정) 미션 제출합니다.

bunju20 opened the pull request:
### 안녕하세요🍀 우아한테크코스 7기 분주입니다!

이번엔 자동차경주라는 과제에서 페어분과 같이 개발을 진행했습니다. 같이 진행하다보니 항상 생각해왔던 방식이 아니라 새로운 방식, 생각하지 못하던 방식들로 개발할수있었어서 무척 의미있었던것 같아요! 개발하면서 궁금했던 점을 적어놨는데, 피드백 해주시면 정말 감사하겠습니다!


### 1. 기능목록
- 입력 : `경주할 자동차 이름을 입력하세요(이름은 쉼표(,)를 기준으로 구분).`
  - 쉼표로 구분하여 배열에 저장
  - 유효성 검사
    - 5자 이하
    - 쉼표를 기준으로 구분
  - (예외) 5자 초과
  - (예외) 0자
  - 실패했을경우 돌아감
- 입력 : `시도할 횟수는 몇 회인가요?`

  - 유효성 검사
    - 양의 정수
  - (예외) 양의 정수가 아닌 값
  - 실패했을경우 돌아감

- 출력 : `실행 결과`
- (반복)현재 게임 상황 출력
  - 각차마다 0부터 9사이의 무작위 값을 구함.
  - 그 값이 4이상이라면 전진한다. 아니라면 정지한다.
  ```
  pobi : -
  woni :
  jun : -
  ```
- 최종우승자 출력
  - 우승자를 쉼표로 나누어 출력(여러명일 경우)
  ```
  최종 우승자: pobi, jun
  ```

### 2. 개발하면서 궁금했던점
* MVC패턴으로 개발을 해야겠다고 생각했는데, 생각보다 시간이 없어서 Controller분리와, model분리를 끝마치지 못한것같아요. 혹시 개발을 진행할때 처음부터 폴더를 분리한뒤에 개발하는건지, 아니면 개발을 하면서 조금씩 분리하는건지 여쭤보고싶습니다!

* 개발하면서 시간복잡도와 가독성 사이에 고민했던 적이 있었어요. 지금은 수정한 코드이지만 프로토타입을 만들때 아래와 같이 가장 큰 값을 구하는 로직이 있었습니다. 사이즈와 난이도는 다르겠지만... 실제에서는 코드 가독성의 우선순위가 시간복잡도보다 높을때가 있는지 여쭤보고싶었어요!

```js
const winnerPosition = [...positionList].sort().reverse().at(0);
```
* OutputView로 분리해야하는게 어느정도인가?를 고민했었습니다. 예시로 아래와 같은 코드가 있으면 이 코드는 어떤 부분을 output으로 분리하는게 적절할까요...?
 `console.log(${nameList[i]} : ${carOutput});` 이부분만 넣으면 오히려 코드를 볼때 이해하기 어려워질것같다고 생각했는데 안분리하자니 책임이 분리되지 않는것 같아 머뭇거리게 된것같아요. 

```js
function printOneGame(nameList, cars) {
  for (let i = 0; i < nameList.length; i++) {
    const carOutput = '-'.repeat(cars[i].position);
    console.log(`${nameList[i]} : ${carOutput}`);
  }
  console.log('');
}

```

긴 글 읽어주셔서 감사합니다!

guesung committed:
docs: 기능 요구사항 작성

guesung committed:
feat: 기능 요구사항 MVP

guesung committed:
feat: 유효성 검사

guesung committed:
feat: 실패 시 다시 시도

guesung committed:
test

guesung committed:
refactor: InputView 분리

guesung committed:
refactor: 유틸 분리

guesung committed:
refactor: 차 모델 분리

guesung committed:
test: 애플리케이션 테스트 코드 작성

guesung committed:
test: 차 모델 테스트 코드 작성

guesung committed:
refactor: 불필요한 파일 제거

swon3210 commented on src/index.js:
```diff
@@ -0,0 +1,50 @@
+  console.log(`최종 우승자: ${winnerOutput}`);
+}
+
+// await run();
```
run 실행문이 주석처리 되어 있으므로 npm run start 를 수행해도 앱이 실행되지 않습니다. 주석을 해제해주세요.

bunju20 replied:
헉... 제가 다시한번 확인하고 요청했어야 했는데 번거롭게 해드려서 죄송합니다ㅠㅠ 알려주셔서 감사합니다!

swon3210 commented on src/utils.js:
```diff
@@ -0,0 +1,35 @@
+  } catch {
+    return await retryUntilSuccess(callbackFn);
+  }
+}
```
[ 첫번째 입력에 공백이 있는 경우 ]

<img width="464" alt="스크린샷 2025-02-13 오후 11 28 23" src="https://github.com/user-attachments/assets/74e57cb2-7abd-4b34-a6b5-78179d88ec42" />


[ 첫번째 입력에 공백 제거 ]

<img width="478" alt="스크린샷 2025-02-13 오후 11 30 21" src="https://github.com/user-attachments/assets/85f91f31-1031-4b35-bb61-21087ba45ea3" />

이름 입력에 공백이 들어간 경우 에러가 나서 정상적으로 입력이 되지 않습니다. 입력에 공백을 허용하지 않는 것을 어플리케이션이 제한하는 것은 가능하겠지만, 사용자가 왜 방금의 입력이 잘못된 것인지를 인지하지 못하는 것은 문제입니다.

```suggestion
export async function retryUntilSuccess(callbackFn) {
  try {
    return await callbackFn();
  } catch {
    console.log('잘못된 입력입니다! 이유 : ~~~~');
  
    return await retryUntilSuccess(callbackFn);
  }
}
```
위와 같이 오류의 이유를 피드백하는 플로우가 필요해보입니다.

또한 한대의 차 이름만을 명시하는게 가능한 상황입니다. 자동차 '경주' 앱인 만큼 2개 이상의 차가 입력되어야 경주 기능이 동작하도록 예외처리 하는 것이 좋아보입니다.

bunju20 replied:
확인해봤습니다! 공백자체를 문자로 인식해서 test1에 띄어쓰기를 하면 공백까지 문자로 인식해서 6자가 넘어 글자수 오류에 걸린거였습니다...!

에러메세지가 있었으면 어떤 오류인지 바로 파악했을것같아서 반성하면서 코드 고치도록 하겠습니다...ㅎㅎ 자동차 갯수도 한개라면 동작하지 않도록 수정하겠습니다. 감사합니다!

swon3210 commented on src/index.js:
```diff
@@ -0,0 +1,50 @@
+    printOneGame(nameList, cars);
+  }
+
+  let winnerPosition = 0;
+  cars.forEach(car => {
+    winnerPosition = Math.max(car.position, winnerPosition);
+  });
```
다음처럼 간소화할 수. 있습니다. 선언과 정의에 필요한 연산을 한번에 하면 let 을 쓸 일이 줄어듭니다.

```suggestion
  const winnerPosition = cars.reduce((maxPosition, car) => 
    Math.max(car.position, maxPosition), 0);
```

bunju20 replied:
훨씬 깔끔한것같습니다...! 항상 사용하던것만 사용하는 문제가 있었는데, 이번기회에 알고있는 지식을 늘릴 필요성을 느낀것같습니다. 피드백 감사합니다!

swon3210 commented on src/index.js:
```diff
@@ -0,0 +1,50 @@
+    winnerPosition = Math.max(car.position, winnerPosition);
+  });
+
+  const winnerList = [];
+  for (let i = 0; i < nameList.length; i++) {
+    const position = cars[i].position;
+    if (position === winnerPosition) winnerList.push(nameList[i]);
+  }
```
아래와 같이 간소화할 수 있습니다.

```suggestion
  const winnerList = cars.filter(car => car.position === winnerPosition);
```

bunju20 replied:
훨씬 직관적이고 이해하기도 쉬운 코드가 된것같아요... 이부분 뿐만아니라 앞으로도 적절한 문법들 사용해서 더 보기 좋게 만들기 명심하겠습니다!

swon3210 commented on src/index.js:
```diff
@@ -0,0 +1,50 @@
+
+// 입출력 예시
+export async function run() {
+  const cars = [];
+
+  const nameList = await InputView.getNameList();
+  const count = await InputView.getCount();
+
+  nameList.forEach(name => {
+    const car = new CarModel(name);
+    cars.push(car);
+  });
```
다음과 같이 간소화 할 수 있습니다.

```suggestion
  const nameList = await InputView.getNameList();
  const count = await InputView.getCount();

  const cars = nameList.map(name => new CarModel(name));
```

bunju20 replied:
cars가 앞에있고 뒤에 처리하는 로직이 있으니 훨씬더 직관적으로 보여서 좋은것 같습니다. 감사합니다...!

swon3210 commented on src/index.js:
```diff
@@ -0,0 +1,50 @@
+  });
+
+  console.log('\n실행 결과');
+  for (let i = 0; i < count; i++) {
+    for (let j = 0; j < nameList.length; j++) {
+      const currentCar = cars[j];
+      currentCar.go();
+    }
+    printOneGame(nameList, cars);
+  }
```
다음과 같이 간소화할 수 있습니다.

```suggestion
  Array.from({ length: count }, () => {
    cars.forEach(car => car.go());
    printOneGame(nameList, cars);
  });
```
중첩된 for 문의 사용은 코드 가독성을 크게 낮춥니다. 최대한 지양해주세요.

bunju20 replied:
Array.from같은 문법을 활용할 생각을 못했었던게 반성이 됩니다... 가독성을 생각하며 고쳐보도록 하겠습니다! 감사합니다!

bunju20 replied:
개별질문에 대한 답변에 코멘트를 못달아서 여기에 적어요!

코드 하나하나 읽어주시고, 캡쳐까지 해주시면서 동작을 확인해주셔서 너무 감사합니다ㅠㅠ 제가 고민했던 부분을 납득할수있었던 여러 말씀들도 너무 감사드려요...! 프론트는 시간복잡도를 크게 신경쓰지 않는군요!

마지막 질문은 프로젝트에서 입출력을 파일로 분리하고싶은데 분리하면 더 이해가 어려워질것같다는 질문이었습니다! 너무 불친절하게 질문한것같아요 정말 죄송하고 반성하고있습니다...ㅠㅠ 하지만 이것도 첫번째로 말씀하신것처럼 이런걸 분리하기 이전에 주어진 부분에 충실하게 구현해야하는것 같습니다! 상세한 피드백 정말 감사합니다!

swon3210 reviewed (changes requested):
로직에 큰 문제가 있는 것은 아니지만 사용자에 대한 고려를 더 해볼 필요가 있다고 생각합니다. 

현업에서 프론트 개발자가 고난이도의 로직을 작성할 일은 적지만, 이용자 편의를 위해 기존 코드를 고치고, 기획자와 디자이너 및 백엔드 개발자를 설득하는 일은 많습니다. 로직 상의 문제만이 버그가 아니고 유저 편의성 관점에서의 불편 또한 버그로 보시는게 좋습니다.

swon3210 commented:
[ 개별 질문에 대한 답변 ]

1. MVC패턴으로 개발을 해야겠다고 생각했는데, 생각보다 시간이 없어서 Controller분리와, model분리를 끝마치지 못한것같아요. 혹시 개발을 진행할때 처음부터 폴더를 분리한뒤에 개발하는건지, 아니면 개발을 하면서 조금씩 분리하는건지 여쭤보고싶습니다!

- 기본적인 폴더 구조와 설계는 프로젝트 시작 시점에 잡아두고 개발을 진행하게 됩니다. 하지만 그것이 모든 코드가 기존의 폴더 구조와 설계를 따라야 함을 의미하지는 않습니다. 또한 현업에서는 우테코 과제를 할 때보다 시간이 부족한 경우가 많습니다. 기존의 구조를 따르는 것이 작업 효율성을 향상시킨다면 따르지만, 그렇지 않은 경우라면 필요한 로직을 작성하고 그것들을 나중에 리팩터링하거나, 더 나은 구조를 위한 논의를 위해 응집해두고 PR 을 올립니다.

- 서비스 구조를 잡는 가장 큰 이유는, 그것이 개발 효율성을 저하시키는 것이 아닌, 촉진시키기 때문이라고 생각합니다. 특정한 구조나 설계를 지키기 위해서 시간이 더 걸렸다고 스스로 평가하신다면, 그러한 구조나 설계가 실제로 프로젝트에 적합한 것인지를 되돌아보셔야 합니다. 제 생각에, 이번 과제의 요구사항을 맞추기 위해 MVC 모델을 도입할 필요는 없었습니다. View 와 Model 을 따로 분리해야할 만큼의 복잡도가 있지 않고 입력 - 출력의 단순 구조를 지닌 어플리케이션이기 때문입니다.


2. 개발하면서 시간복잡도와 가독성 사이에 고민했던 적이 있었어요. 지금은 수정한 코드이지만 프로토타입을 만들때 아래와 같이 가장 큰 값을 구하는 로직이 있었습니다. 사이즈와 난이도는 다르겠지만... 실제에서는 코드 가독성의 우선순위가 시간복잡도보다 높을때가 있는지 여쭤보고싶었어요!

- 사견이지만, 프론트엔드 개발자는 시간복잡도에 대해 크게 신경쓰지 않아도 됩니다. 거의 모든 자바스크립트 코드는 밀리세컨드 단위로 연산이 완료됩니다. 0.001 초와 0.1 초의 차이는 100배이지만 이용자는 체감하지 못할 가능성이 큽니다. 성능을 위해 가독성을 희생할 바에는 성능을 희생하고 가독성을 챙기는 것이 훨씬 나은 선택입니다. 그 반대가 성립될 수 있는 요건은 '이용자의 편의성'에 차이가 생기는가의 여부입니다.

- 백엔드는 시간복잡도가 중요합니다. 연산에 회사의 컴퓨팅 자원을 쓰기 때문입니다. 따라서 연산을 효율적으로 만들 수록 비용을 줄이고 서비스 품질도 향상시킬 수 있습니다. 그러나 프론트엔드는 대체로 이용자 기기의 컴퓨팅 자원을 사용하여 연산합니다. 연산을 효율화 시킨다해도 비용과 품질 측면에서 큰 차이가 없을 가능성이 큽니다.

3. OutputView로 분리해야하는게 어느정도인가?를 고민했었습니다. 예시로 아래와 같은 코드가 있으면 이 코드는 어떤 부분을 output으로 분리하는게 적절할까요...?
console.log(${nameList[i]} : ${carOutput}); 이부분만 넣으면 오히려 코드를 볼때 이해하기 어려워질것같다고 생각했는데 안분리하자니 책임이 분리되지 않는것 같아 머뭇거리게 된것같아요.

- 질문을 제대로 이해하지 못했습니다. 'OutputView' 라는게 정확히 무엇을 의미하는 것인가요?

bunju20 committed:
Update src/utils.js

Co-authored-by: SONG WON LEE <swon3210@gmail.com>

bunju20 committed:
Update src/index.js

Co-authored-by: SONG WON LEE <swon3210@gmail.com>

bunju20 committed:
Update src/index.js

Co-authored-by: SONG WON LEE <swon3210@gmail.com>

bunju20 committed:
Update src/index.js

Co-authored-by: SONG WON LEE <swon3210@gmail.com>

bunju20 committed:
Update src/index.js

Co-authored-by: SONG WON LEE <swon3210@gmail.com>

bunju20 committed:
fix: refactor : 오류메세지 출력 및 예외처리 추가
- run 주석처리 제거함
- 오류 메세지 이유 출력하도록 수정
- 자동차 입력시 공백이 있을때 공백을 제거하는 로직 추가
- 차 한개가 입력됐을때 오류가 나도록 예외처리

bunju20 committed:
fix: 병합 충돌 수정:

bunju20 committed:
fix: 최종우승자 출력 오류 수정

bunju20 committed:
fix: console.log 출력 삭제

bunju20 committed:
fix: 우승자 출력시 공백 제거

swon3210 reviewed (approved):
안녕하세요 분주! 저야말로 리뷰해드린 내용 성실히 반영해주시고 답변 주셔서 감사드립니다. 마지막 질문은 제가 이해력이 부족해서 이해가 안된다고 말씀드렸을 뿐이니 죄송해하지 않으셔도 됩니다! 입출력 로직을 파일로 분리하는 것이 좋은지에 대한 질문이셨군요! 저는 지금 요구사항에 따르면 지금의 코드로도 충분히 코드를 이해하는데 문제가 없기에 굳이 그럴 필요까지는 없다고 생각합니다!

제가 드린 리뷰는 '생산성'에 가장 큰 중점을 두고 있는 저의 견해를 바탕으로 작성된 것이기에, 분주나 다른 리뷰어들과는 생각이 조금씩 다를 수 있습니다! 가능하다면 여러 견해들을 접해보시면서 최대한 많은 코드를 작성해보시길 권장드립니다.

미션 진행하시느라 수고많으셨습니다!
